### PR TITLE
Add neo client config struct & ctor

### PR DESCRIPTION
**Pull request by gavin-norman-sociomantic**

### CODE REVIEW

**Comment by gavin-norman-sociomantic** [comment]
Will add a test case reading the `Config` instance from a file.

**Comment by gavin-norman-sociomantic** [comment]
Updated with test case.

**Comment by scott-gibson-sociomantic** [comment]
I can't speak for other team's apps but i know that reporting apps take advantage of the previous class based IClient.Config by inheriting and adding things such as throttle information, channels to read etc. 
Do you want to limit that behaviour here and prefer apps to separate the config?

**Comment by gavin-norman-sociomantic** [comment]
Interesting. I went for a struct as I couldn't imagine why anyone would need to inherit the config container. Tbh it sounds to me like what you describe is... kind of weird...? I mean, throttling is a separate issue to the basic configuration of a swarm client.

That said, my intention wasn't to deliberately limit usage, so maybe I should change it to a class.

**Comment by gavin-norman-sociomantic** [comment]
Updated. Config is now a class.

**Comment by gavin-norman-sociomantic** [comment]
>Interesting. I went for a struct as I couldn't imagine why anyone would need to inherit the config container.

And we consciously avoid classes and inheritance, generally, in neo.

**Comment by scott-gibson-sociomantic** [comment]
`Interesting. I went for a struct as I couldn't imagine why anyone would need to inherit the config container.`

Legacy does this itself with the Scheduling*Client config. I'm not sure how neo approaches this sort of behaviour but it's worth a thought.

**Comment by scott-gibson-sociomantic** [comment]
For example in TubeEye,  i have my dmq input class taking a config instance of
```
public class Dmq : SchedulingDmqClient.Config
{
    /// The Dmq channel to output processed records to.
    public ConfigFiller.Required!(char[]) channel;
}
```

**Comment by gavin-norman-sociomantic** [comment]
>Legacy does this itself with the Scheduling*Client config. I'm not sure how neo approaches this sort of behaviour but it's worth a thought.

Right, this was necessary internally in swarm with the legacy client. This is not necessary with the neo client and is indeed shunned, as I mentioned above. That's why I thought it was unnecessary.

**Comment by scott-gibson-sociomantic** [comment]
Again, i would like some input from the other teams on if they do a similar thing. It's hard to search github :(

**Comment by gavin-norman-sociomantic** [comment]
Yep. It does no harm being a class, though, so whether a certain usage is categorised by me as "weird" or not doesn't really matter ;)

File: src/swarm/neo/client/mixins/ClientCore.d [hunk 1]
```diff
@@ -92,6 +92,24 @@ template ClientCore ( )
 
     /***************************************************************************
 
+        Config struct which may be passed to the ctor. Designed for use with
+        ocean's ConfigFiller helper.
+
+    ***************************************************************************/
+
+    public static class Config
```
Comment (nemanja-boric-sociomantic) [inline src/swarm/neo/client/mixins/ClientCore.d:100]: commit message is outdated, this is a `class`.

File: src/swarm/neo/client/mixins/ClientCore.d [hunk 2]
```diff
@@ -92,6 +92,24 @@ template ClientCore ( )
 
     /***************************************************************************
 
+        Config struct which may be passed to the ctor. Designed for use with
```
Comment (nemanja-boric-sociomantic) [inline src/swarm/neo/client/mixins/ClientCore.d]: `struct`

File: relnotes/config.feature.md [hunk 3]
```diff
@@ -0,0 +1,6 @@
+* `swarm.neo.client.mixins.ClientCore`
+
+  A new constructor has been added that accepts a ``Config`` struct instance
```
Comment (nemanja-boric-sociomantic) [inline relnotes/config.feature.md]: `s/struct/class` in multiple places I guess.

**Comment by nemanja-boric-sociomantic** [comment]
Generally, LGTM

File: relnotes/config.feature.md (hunk 3 above)
Comment (gavin-norman-sociomantic) [reply relnotes/config.feature.md]: Updated.

Comment (gavin-norman-sociomantic) [reply relnotes/config.feature.md]: Oh wait, I missed the commit message.

Comment (gavin-norman-sociomantic) [reply relnotes/config.feature.md]: Updated.

Comment (nemanja-boric-sociomantic) [reply relnotes/config.feature.md]: This is still `Config` struct instance.

Comment (gavin-norman-sociomantic) [reply relnotes/config.feature.md]: Ha I missed this one completely ;)

Comment (gavin-norman-sociomantic) [reply relnotes/config.feature.md]: Updated again.

**Pull request merged**